### PR TITLE
Add some new operations to `AnnotatedMap` and reimplement `BoolMap`

### PR DESCRIPTION
**Pull request by robdockins**
in terms of it.  This should prevent quatradic-time behavior of
`BoolMap` due to hashing.

### CODE REVIEW

**Comment by robdockins** [comment]
@brianhuffman, I added a bunch more since you took a look earlier.

File: what4/src/What4/Expr/Builder.hs [hunk 1]
```diff
@@ -2860,9 +2816,7 @@ sbMakeExpr sym a = do
     modifyIORef' (sbNonLinearOps sym) (+1)
   case appType a of
     -- Check if abstract interpretation concludes this is a constant.
-    BaseBoolRepr | Just b <- v -> return $ backendPred sym b
```
Comment (robdockins) [inline what4/src/What4/Expr/Builder.hs]: On second thought, I think we'd better put this line back in... even though the boolean operations don't really compute much, this value can still matter when inserting/projecting from structures and arrays.

File: what4/src/What4/Expr/WeightedSum.hs [hunk 2]
```diff
@@ -574,36 +621,43 @@ extractCommon (WeightedSum xm xc sr) (WeightedSum ym yc _) = (z, x', y')
...
+prodVar sr x = mkProd sr (singletonProdMap sr (SR.occ_one sr) x)
 
 -- | Multiply two products, collecting terms and adding occurences.
 prodMul :: Tm f => SemiRingProduct f sr -> SemiRingProduct f sr -> SemiRingProduct f sr
 prodMul x y = mkProd sr m
   where
   sr = prodRepr x
-  mergeCommon _ a b = Just (SR.occ_add sr a b)
-  m = Map.mergeWithKey mergeCommon id id (_prodMap x) (_prodMap y)
+  mergeCommon _ (u,a) (v,b) = Just (u <> v, SR.occ_add sr a b)
```
Comment (robdockins) [inline what4/src/What4/Expr/WeightedSum.hs]: Oh... pretty sure this isn't right, actually.  This will get the wrong hash value

Comment (brianhuffman) [reply what4/src/What4/Expr/WeightedSum.hs]: Yeah, you're going to want to call `mkProdNote` there, just like `mkNote` is used in the implementation of function `add`

**Comment by brianhuffman** [comment]
I think this is ready to merge; @robdockins, are we all set to go?

**Comment by robdockins** [comment]
I was just waiting for your thumbs up!

**Pull request merged**